src/lib/dict.c

```diff
@@ -0,0 +1,142 @@
+#include <rlang.h>
+#include "dict.h"
+
+#define DICT_LOAD_THRESHOLD 0.75
+#define DICT_GROWTH_FACTOR 2
+
+static size_t size_round_power_2(size_t size);
+
+
+struct r_dict r_new_dict(r_ssize size) {
+  if (size <= 0) {
+    r_abort("`size` of dictionary must be positive.");
+  }
+  size = size_round_power_2(size);
```

Comment:
Size can't be negative here so this should be safe. We don't check for overflow though. This may be overkill since `size_t` should be extremely large on 64 bits platforms:

```c
static
size_t size_round_power_2(size_t size) {
  size_t out = 1;
  size_t last = SIZE_T_MAX;

  while (out < size) {
    out <<= 1;

    // Protect from wrap-around
    if (out < last) {
      r_abort("Integer overflow in `size_round_power_2()`.");
    }
    last = out;
  }

  return out;
}
```

Then we also need to check for `r_ssize` overflow. This could only be a problem when R is compiled without long vector support since `r_ssize` is then an `int`. Even so, I don't expect this to go over 4 billions though.

Reply:
I imagine that what you already have is probably fine